theseus/geometry/functional/so3.py

```diff
@@ -376,3 +389,125 @@ def backward(cls, ctx, grad_output):
 _jcompose_autograd_fn = _jcompose_impl
 
 compose, jcompose = lie_group.BinaryOperatorFactory(_module, "compose")
+
+
+# -----------------------------------------------------------------------------
+# Unit Quaternion to Rotation Matrix
+# -----------------------------------------------------------------------------
+def _quaternion_to_rotation_impl(quaternion: torch.Tensor) -> torch.Tensor:
+    if quaternion.ndim == 1:
+        quaternion = quaternion.unsqueeze(0)
+    check_unit_quaternion(quaternion)
+
+    quaternion = quaternion / torch.norm(quaternion, dim=1, keepdim=True)
+    w = quaternion[:, 0]
+    x = quaternion[:, 1]
+    y = quaternion[:, 2]
+    z = quaternion[:, 3]
+
+    q00 = w * w
+    q01 = w * x
+    q02 = w * y
+    q03 = w * z
+    q11 = x * x
+    q12 = x * y
+    q13 = x * z
+    q22 = y * y
+    q23 = y * z
+    q33 = z * z
+
+    ret = quaternion.new_zeros(quaternion.shape[0], 3, 3)
+    ret[:, 0, 0] = q00 + q11 - q22 - q33
+    ret[:, 0, 1] = 2 * (q12 - q03)
+    ret[:, 0, 2] = 2 * (q13 + q02)
+    ret[:, 1, 0] = 2 * (q12 + q03)
+    ret[:, 1, 1] = q00 - q11 + q22 - q33
+    ret[:, 1, 2] = 2 * (q23 - q01)
+    ret[:, 2, 0] = 2 * (q13 - q02)
+    ret[:, 2, 1] = 2 * (q23 + q01)
+    ret[:, 2, 2] = q00 - q11 - q22 + q33
+    return ret
+
+
+def _jquaternion_to_rotation_impl(
+    quaternion: torch.Tensor,
+) -> Tuple[List[torch.Tensor], torch.Tensor]:
+    if quaternion.ndim == 1:
+        quaternion = quaternion.unsqueeze(0)
+    check_unit_quaternion(quaternion)
+
+    quaternion_norm = torch.norm(quaternion, dim=1, keepdim=True)
+    quaternion = quaternion / quaternion_norm
+    w = quaternion[:, 0]
+    x = quaternion[:, 1]
+    y = quaternion[:, 2]
+    z = quaternion[:, 3]
+
+    q00 = w * w
+    q01 = w * x
+    q02 = w * y
+    q03 = w * z
+    q11 = x * x
+    q12 = x * y
+    q13 = x * z
+    q22 = y * y
+    q23 = y * z
+    q33 = z * z
+
+    ret = quaternion.new_zeros(quaternion.shape[0], 3, 3)
+    ret[:, 0, 0] = q00 + q11 - q22 - q33
+    ret[:, 0, 1] = 2 * (q12 - q03)
+    ret[:, 0, 2] = 2 * (q13 + q02)
+    ret[:, 1, 0] = 2 * (q12 + q03)
+    ret[:, 1, 1] = q00 - q11 + q22 - q33
+    ret[:, 1, 2] = 2 * (q23 - q01)
+    ret[:, 2, 0] = 2 * (q13 - q02)
+    ret[:, 2, 1] = 2 * (q23 + q01)
+    ret[:, 2, 2] = q00 - q11 - q22 + q33
+
+    temp = -2 * quaternion / quaternion_norm
+    jac = quaternion.new_zeros(quaternion.shape[0], 3, 4)
+    jac[:, :, :1] = temp[:, 1:].view(-1, 3, 1)
+    jac[:, :, 1:] = hat(temp[:, 1:])
+    jac[:, 0, 1] = -temp[:, 0]
+    jac[:, 1, 2] = -temp[:, 0]
+    jac[:, 2, 3] = -temp[:, 0]
+
+    return [jac], ret
+
+
+class QuaternionToRotation(lie_group.UnaryOperator):
```

Comment:
Will we add rotation to quaternion later? From our internal doc there are potential users who would like support for quaternions as well as other representations like axis-angle. Not necessary for this PR but we should make sure we are setting up the API in a way that adding those later is easier.